server/server.go

```diff
@@ -277,18 +286,28 @@ func RegisterInstrumentation(router *mux.Router) {
 	router.PathPrefix("/debug/pprof").Handler(http.DefaultServeMux)
 }
 
-// Run the server; blocks until SIGTERM or an error is received.
+// Run the server; blocks until SIGTERM (if signal handling is enabled), an error is received, or Stop() is called.
 func (s *Server) Run() error {
 	errChan := make(chan error, 1)
 
-	// Wait for a signal
-	go func() {
-		s.handler.Loop()
-		select {
-		case errChan <- nil:
-		default:
-		}
-	}()
+	if s.handler != nil {
+		// Wait for a signal
+		go func() {
+			s.handler.Loop()
+			select {
+			case errChan <- nil:
+			default:
+			}
+		}()
+	} else {
+		go func() {
+			<-s.quit
```

Comment:
Seems like we have many ways to quit now and I don't know why.
Why do we need Stop vs Shutdown?

Reply:
`Stop` vs `Shutdown` was there before. `Run` is a blocking method, that returns first error from HTTP/gRPC or nil, if signal has been received or `Stop` was called.
`Shutdown` stops gRPC and HTTP servers.

Since `Stop` works by stopping signal handler, which can no longer be there, I've kept the functionality of `Stop` by using `quit` channel.

If we were doing design from scratch, we could replace blocking `Run` method with non-blocking `Start` (perhaps returning channel with errors), which would make `Stop` unnecessary.

Reply:
Looking further, the signal-handler has _another_ `Stop()` mechanism `SignalReceiver`, which appears to be impossible to use from `Server`.

How about making `handler` something you can pass in config, like `Log`, that has interface `Loop()` and `Stop()`, then Cortex can pass either what you're doing with `quit` or its own actual signal handler?

Reply:
That sounds good, I'll update PR with your suggestion.